pkg/dev/dev.go

```diff
@@ -74,3 +96,107 @@ func (o *DevClient) Cleanup() error {
+	var err error
+	var containers []v1.Component
+	containers, err = devfileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1.ContainerComponentType},
+	})
+
+	endpoints := endpointsFromContainers(containers)
+	portPairs := portPairsFromEndpoints(endpoints)
+
+	if len(portPairs) == 0 {
+		// no endpoints with exposure set to public or internal; no ports to be forwarded
+		return nil
+	}
+
+	var pod *corev1.Pod
+	pod, err = o.kubernetesClient.GetOnePodFromSelector(componentlabels.GetSelector(devfileObj.GetMetadataName(), envSpecificInfo.GetApplication()))
+	if err != nil {
+		return err
+	}
+
+	transport, upgrader, err := spdy.RoundTripperFor(o.kubernetesClient.GetClientConfig())
+	if err != nil {
+		return err
+	}
+
+	req := o.kubernetesClient.GeneratePortForwardReq(pod.Name)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	stopChan := make(chan struct{}, 1)
+	// passing nil for readyChan because it's eventually being closed if it's not nil
+	// passing nil for out because we only care for error, not for output messages; we want to print our own messages
+	var fw *portforward.PortForwarder
+	fw, err = portforward.NewOnAddresses(dialer, []string{"localhost"}, portPairs, stopChan, nil, nil, errOut)
+	if err != nil {
+		return err
+	}
+
+	// start port-forwarding
+	go func() {
+		err = fw.ForwardPorts()
+		if err != nil {
+			fmt.Fprint(out, fmt.Errorf("error setting up port forwarding: %v", err).Error())
+			os.Exit(1)
+		}
+	}()
+
+	printPortForwardingInfo(portPairs, out)
+	return nil
+}
+
+func printPortForwardingInfo(portPairs []string, out io.Writer) {
+	portFowardURLs := "You can access "
+	for i := range portPairs {
+		split := strings.Split(portPairs[i], ":")
+		local := split[0]
+		remote := split[1]
+
+		if i == len(portPairs)-1 && i != 0 {
+			portFowardURLs += fmt.Sprintf("and port %s at http://localhost:%s", remote, local)
+			break
+		} else if i < len(portPairs)-2 {
+			portFowardURLs += fmt.Sprintf("port %s at http://localhost:%s, ", remote, local)
+		} else {
+			portFowardURLs += fmt.Sprintf("port %s at http://localhost:%s ", remote, local)
+		}
+	}
+	fmt.Fprintf(out, "\n%s", portFowardURLs)
+}
+
+// endpointsFromContainers returns a slice of endpoints with exposure value set to public or internal
+func endpointsFromContainers(containers []v1alpha2.Component) []v1.Endpoint {
+	var endpoints []v1.Endpoint
+
+	for i := range containers {
+		for _, ep := range containers[i].Container.Endpoints {
+			if ep.Exposure != v1.NoneEndpointExposure {
+				endpoints = append(endpoints, ep)
+			}
+		}
+	}
+
+	return endpoints
+}
+
+// portPairsFromEndpoints returns a slice of strings of the format "<local-port>:<remote-port>"
+func portPairsFromEndpoints(endpoints []v1alpha2.Endpoint) []string {
+	var portPairs []string
+	var port = 40000
+
+	for _, e := range endpoints {
+	loop:
```

Comment:
You could do this with a `for` loop and using `continue` instead of `goto loop` and a break when the port is found

Reply:
That would require a nested for loop, which I was trying to avoid here.

Reply:
But you are also creating a loop with this label/goto. And IMO, it is less easy to read than a `for` loop

If you have 2 dimensions to explore, I cannot see the problem to have a nested loop here